File: ios/Services/KitharaPlayer/Sources/ServicesKitharaPlayer/AudioPlaybackEngine.swift
import Foundation

protocol AudioPlaybackEngine: Sendable {
    var currentState: AudioEngineState { get }

    func engineStateStream() -> AsyncStream<AudioEngineState>
    func eventStream() -> AsyncStream<AudioEngineEvent>

    func play()
    func pause()
    func seekTo(positionMs: Int64) async -> Bool
    func setQueueWindow(current: AudioTrackRequest, next: AudioTrackRequest?, autoPlay: Bool)
    func appendNext(_ next: AudioTrackRequest)
    func selectInWindow(appItemId: String, autoPlay: Bool) async -> Bool
    func pruneWindow(keepAppItemIds: Set<String>)
    func stop()
}

File: android/services/kithara/src/main/java/com/mplayeraudio/services/kithara/AudioPlaybackEngine.kt
package com.mplayeraudio.services.kithara

import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow

interface AudioPlaybackEngine {
    val engineState: StateFlow<AudioEngineState>
    val events: SharedFlow<AudioEngineEvent>

    fun play()
    fun pause()
    suspend fun seekTo(positionMs: Long): Boolean
    fun setQueueWindow(
        current: AudioTrackRequest,
        next: AudioTrackRequest?,
        autoPlay: Boolean,
    )

    fun appendNext(next: AudioTrackRequest)

    suspend fun selectInWindow(appItemId: String, autoPlay: Boolean): Boolean

    fun pruneWindow(keepAppItemIds: Set<String>)

    fun stop()
}
